### ISSUE TITLE
UX improvements for CFE error messages

### ISSUE THREAD

**Issue by mraleph** [dart-lang/sdk#50281]
When reporting errors CFE (or front-end tools that wrap CFE) tend to mix errors from the entry point, entry-point package and its dependencies into a single output. This becomes extremely confusing (especially to the new comers) when errors occur somewhere in the `.pub_cache`. The most common sources of confusion:

* Importing Web library on the Native (and vice versa), e.g. https://github.com/dart-lang/sdk/issues/47260
* Hitting a breaking change which prevents dependencies from compiling, e.g. https://github.com/flutter/flutter/issues/113575
* `.pub_cache` corruption due to accidental or stale edits.

Here is a verbatim example from the second case: 

```
../../../../../.pub-cache/hosted/pub.dartlang.org/file-6.1.2/lib/src/interface/file.dart:15:16: Error: The method 'File.create' has fewer named arguments than those of overridden method 'File.create'.
  Future<File> create({bool recursive = false});
               ^
org-dartlang-sdk:///third_party/dart/sdk/lib/io/file.dart:232:16: Context: This is the overridden method ('create').
  Future<File> create({bool recursive = false, bool exclusive = false});
               ^
../../../../../.pub-cache/hosted/pub.dartlang.org/file-6.1.2/lib/src/interface/file.dart:15:16: Error: The method 'File.create' doesn't have the named parameter 'exclusive' of overridden method 'File.create'.
  Future<File> create({bool recursive = false});
               ^
org-dartlang-sdk:///third_party/dart/sdk/lib/io/file.dart:232:16: Context: This is the overridden method ('create').
  Future<File> create({bool recursive = false, bool exclusive = false});
               ^
../../../../../.pub-cache/hosted/pub.dartlang.org/file-6.1.2/lib/src/forwarding/forwarding_file.dart:22:8: Error: The method 'ForwardingFile.createSync' has fewer named arguments than those of overridden method 'File.createSync'.
  void createSync({bool recursive = false}) =>
       ^
org-dartlang-sdk:///third_party/dart/sdk/lib/io/file.dart:248:8: Context: This is the overridden method ('createSync').
  void createSync({bool recursive = false, bool exclusive = false});
       ^
../../../../../.pub-cache/hosted/pub.dartlang.org/file-6.1.2/lib/src/forwarding/forwarding_file.dart:22:8: Error: The method 'ForwardingFile.createSync' doesn't have the named parameter 'exclusive' of overridden method 'File.createSync'.
  void createSync({bool recursive = false}) =>
       ^
org-dartlang-sdk:///third_party/dart/sdk/lib/io/file.dart:248:8: Context: This is the overridden method ('createSync').
  void createSync({bool recursive = false, bool exclusive = false});
```

There are multiple sources of confusion here. SDK uris are not actionable for a novice. `../../../../../.pub-cache/hosted/pub.dartlang.org/file-6.1.2` is just visual noise. 

I propose we make few adjustments in how CFE reports the errors (by default, possibly with escape hatch for advanced users). 

* Dart SDK locations should be shortened to something like `Dart SDK` as they are non actionable for a user anyway.
* Errors should be visually grouped by packages and potentially fully hidden (with a mode to unhide them). 
* Some suggestions on how to address the issue (e.g. by upgrading dependencies or fixing pub cache) should be made.
* There should be a mode which shows an import path towards some library with an error. 

This means we should produce something like this instead of an error above: 

```
Compilation errors in package 'file' (version 6.1.2):

  package:file/src/interface/file.dart:15:16: Error: The method 'File.create' has fewer named arguments than those of overridden method 'File.create'.
    Future<File> create({bool recursive = false});
                 ^
  Dart SDK: Context: This is the overridden method ('create').
    Future<File> create({bool recursive = false, bool exclusive = false});
                 ^
  package:file/src/interface/file.dart:15:16: Error: The method 'File.create' doesn't have the named parameter 'exclusive' of overridden method 'File.create'.
    Future<File> create({bool recursive = false});
                 ^
  Dart SDK: Context: This is the overridden method ('create').
    Future<File> create({bool recursive = false, bool exclusive = false});
                 ^
  package:file/src/forwarding/forwarding_file.dart:22:8: Error: The method 'ForwardingFile.createSync' has fewer named arguments than those of overridden method 'File.createSync'.
    void createSync({bool recursive = false}) =>
         ^
  Dart SDK: Context: This is the overridden method ('createSync').
    void createSync({bool recursive = false, bool exclusive = false});
         ^
  package:file/src/forwarding/forwarding_file.dart:22:8: Error: The method 'ForwardingFile.createSync' doesn't have the named parameter 'exclusive' of overridden method 'File.createSync'.
    void createSync({bool recursive = false}) =>
         ^
  Dart SDK: Context: This is the overridden method ('createSync').
    void createSync({bool recursive = false, bool exclusive = false});

       ^
The package version currently in use might be incompatible with other 
dependencies or current Flutter SDK version. Try upgrading package file
by running: 

    $ pub ... 
```

another option is to fully collapse these errors away

```
Error: compilation errors occurred when building libraries belonging to package 'file' (version 6.1.2). 
          Run with -v to see them. The package version currently in use might be incompatible with other
          dependencies or current Flutter SDK version. Try upgrading package file by running: pub ... 
```

(At some point `pub` team is going to provide package checksums so we should be able to detect dirty `.pub_cache` and issue appropriate warning as well)

/cc @jensjoha @johnniwinther @chloestefantsova 
/cc @mit-mit

**Comment by mit-mit** [comment]
Yes, please. cc @jacob314 who also had some feedback on this

**Comment by lrhn** [comment]
I completely agree that you should show the URI of the library in most cases, *not* the file path, if the library was imported with a package URI.
For the current package, showing file paths may make sense. (I'm not even sure about that.)
For files actually accessed using a `file:` URI, we can show that.

For Dart SDK links, we should at least have `dart:io` in the line (again, the actual import URI). I'd be fine with using just `dart:io`.
```
 dart:io: Context: This is the overridden method ('createSync').
    void createSync({bool recursive = false, bool exclusive = false});

       ^
```

For my own purposes, I'd be very sorry debugging the SDK if I can't get file and line numbers, I'd prefer
```
 dart:io(file.dart:248:8): Context: This is the overridden method ('createSync').
    void createSync({bool recursive = false, bool exclusive = false});

       ^
```
since that actually does give you something to go on (even if it's just looking at the SDK source, which some developers do), but isn't intrusive if you don't need it.
Otherwise I'll have to ask for a flag to get useful SDK stack trace entries. I'd prefer to only have one code path.

**Comment by jensjoha** [comment]
> * Dart SDK locations should be shortened to something like `Dart SDK` as they are non actionable for a user anyway.

I disagree that it's not actionable. Being able to go and find the source to look at it can be very helpful. "SDK says no" is not. I do agree that "org-dartlang-sdk:///third_party/dart/..." isn't all that helpful, but something like "dart:io/file.dart:232:16" might be (which I think matches what a stacktrace might show). It should then preferably be clickable in the editor which should then load the correct platform source (if run from such a thing).

Similarly I do agree that showing an import uri seems like a better choice than an file uri for packages.

> * Errors should be visually grouped by packages and potentially fully hidden (with a mode to unhide them).

I don't understand this part. Errors should be hidden?
User: Run this.
dart: No.

Update: You actually had an example of this, okay --- but that specifically mentions the "flutter sdk" which hardly seems great either.
Also I'm personally a fan of information, so I don't really like hiding it.
If upgrading doesn't work and an issue is posted, the first question from a team member will be "can you give us the data when run with -v".
If your own code is in a package (as for instance the case for `flutter create foo` or `dart create foo`) you won't see that either...

As for grouping them I'm not sure what I think... In the example it does seem nice, but I'm not sure that generalizes. Also, currently we generally give an error when we see it. Grouping it would certainly mean postponing until we're done (and thus certain that we'll report no more errors). (This will then present the issue of remembering to issue errors, the timing of which depends on whether just doing outlines or doing a full compile).

Asking to run `pub upgrade` or similar is also one-sided (user of a package, not creators of a package). For instance when saying `flutter create foo` it creates a package called `foo` so the suggestion would mean suggesting all developers to fix any compilation bug by running `pub upgrade`.

It also adds a sorting issue. Currently errors are printed as we see them, hopefully by what caused it first. If grouping, what do we print first?

> Compilation errors in package 'file' (version 6.1.2)

We don't have access to that information. Nor do I think it necessarily adds much value.

> * There should be a mode which shows an import path towards some library with an error.

I don't understand this part.

**Comment by jensjoha** [comment]
> When your dependency does not compile on the other hand - you are not going to fix it. You are going to try to upgrade it. Compilation errors in the dependencies usually means that something is going really wrong.

While I might not go and fix it, I am going to take a look at it. (actually, I have changed stuff in packages quite often to debug stuff, though, maybe not to fix a compilation error).
But also: When everything's in a package we can't really know what's you're code and what isn't.

**Comment by stuartmorgan** [comment]
> Asking to run `pub upgrade` or similar is also one-sided (user of a package, not creators of a package). For instance when saying `flutter create foo` it creates a package called `foo` so the suggestion would mean suggesting all developers to fix any compilation bug by running `pub upgrade`.

This seems like something that could be fixed with wording, e.g. "If you aren't the author of package `foo`, you may need to run `{dart,flutter} pub upgrade`, or contact the developers". It's very easy for package developers to ignore a message as long as it's written in such a way that it's clear when it's not applicable, whereas it can be hard for someone with no idea what errors from a package even mean—which we see regularly in the Flutter issue tracker—to know what to do when there's no suggestion.

**Comment by jensjoha** [comment]
Personally I don't think that fixes much. It's just more text.

With the other suggestions it will still hide the compilation errors where they're most likely to occur --- in the users own code.

In the example with "package:js/js.dart", running `flutter pub upgrade` won't help and it's not really the package that's at fault.

On top of that the compiler doesn't know if you're running from `flutter` or from `dart`.

Maybe I should make it clear that I'm not opposed to improving the UX --- I fully support that. I agree that some of these suggestions would improve things (e.g. using import uri instead of file uri). On the other hand I disagree with some of it, where it seems to me we're trying to optimize for a scenario that, while it does occur, occurs much less often than other scenarios where the suggestion will make things worse.

**Comment by mraleph** [comment]
> In the example with "package:js/js.dart", running flutter pub upgrade won't help and it's not really the package that's at fault.

Yeah, CFE should be smarter at classifying the error and then giving an actionable suggestion. 

That being said, let's take it step by step - rather than get stuck on more controversial details. How about we at least implement the following, which all seems uncontroversial: 

- use import uris
- clearly group error messages by package
- shorten SDK uris more readable, when compiling user code (as opposed to compiling platform)
- (stretch goal) for problems with platform imports like `dart:js` prove import chain leading to the import of unsupported library.

This removes controversial things like completely hiding error messages and making concrete suggestions on how to fix errors.

**Comment by jensjoha** [comment]
> * clearly group error messages by package

This one still seems controversial to me, as I've also outlined previously.

> * shorten SDK uris into more readable forms (I like Lasse's suggestion), when compiling user code (as opposed to compiling platform)

This one, I think, is weird. `dart:io(file.dart:248:8)` is a weird way to write anything --- if using the import uri (bullet number 1) it will become (something like) `dart:io/file.dart:248:8` which seems more consistent to me.

**Comment by lrhn** [comment]
I think `dart:io/file.dart:248:8` is perfectly fine. It's a little weird in that it's a URI that users can't write (it's not a valid `dart:` URI, as much as we have ever defined what that is), but for error reporting it contains all the available information in a reasonable format. SGTM!

**Comment by mit-mit** [comment]
`dart:io/file.dart:248:8` sgtm

**Comment by jensjoha** [comment]
> If making suggestions about how to fix errors is controversial at the Dart level

I don't think that's a fair summary.
On the basis of this example, though, I find, that the compiler suggesting to run `flutter pub upgrade` whenever there's a compile time error in a package, would be weird.

Maybe "controversial" is the wrong word though. There were suggestions made that we could agree upon, and suggestions we could not. The ones we cannot probably warrant more discussion.

**Comment by stuartmorgan** [comment]
> > If making suggestions about how to fix errors is controversial at the Dart level
> 
> I don't think that's a fair summary.

To be clear, this wasn't intended to be a blanket statement about Dart, it was a response to the specific comment about the suggestions above being controversial in this context. I understand why the CFE might not want to make that suggestion, but if the CFE could make it easy for Flutter to identify cases where we could make that suggestion, it would be helpful, because I think it would be much more reasonable to add heuristically at that level.

**Comment by jensjoha** [comment]
What if

* Pub added a field to `.dart_tool/package_config.json`, marking stuff from pub, so we could distinguish packages fetched from pub from packages we either wrote ourselves, dir overwrote or whatever.
* Extracted information about the tool so we can distinguish `flutter` and `dart` runs --- possibly this can be done with the `target`, but I'm not sure.
* Experimented with adding a "summary" of sorts where, if there were compilation errors in one or more package from pub, we could offer suggestions (which might be more than `flutter pub upgrade`).

That way we wouldn't hide information, if some path forgot to give the summary it wouldn't be the end of the world, there wouldn't really be a sorting problem, we wouldn't ask the user to `flutter pub upgrade` their way out of their own compilation errors, and only in the weird situations like the example here, would stuff actually change.

Thoughts?

**Comment by devoncarew** [comment]
> Experimented with adding a "summary" of sorts where, if there were compilation errors in one or more package from pub, we could offer suggestions (which might be more than flutter pub upgrade).

This work can be independent of the CFE; we catch and handle the CFE compilation errors in dartdev (https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/src/commands/run.dart#L305), so can improve the what-to-do-on-failures message there.

Shortening the CFE paths - to dart: and package: urls - makes a lot of sense; this would improve the readability of the errors. Note that for the package:file issue that we're hitting, the version of the package is an important part of the error message (6.1.2 doesn't work on newer SDKs). So for:

```
../../../../../.pub-cache/hosted/pub.dartlang.org/file-6.1.2/lib/src/interface/file.dart:15:16
```

it would be great to be able to preserve the package version info when shortening the path. I'm not sure if that means creative regexing of the above path or means that we'd like more metadata in the package_config.json file.

**Comment by mit-mit** [comment]
For hosted packages from pub.dev, maybe something like:
`pub.dev/file:6.1.2/lib/src/interface/file.dart:15:16` ?
